File: XPR/C-Version/XPR-0.C/parser.cpp
/**
 * Este arquivo implementa o reconhecimento sintático (parsing) da
 * linguagem XPR-0, que representar expressões algébricas simples com
 * operações de adição e multiplicação.
 * 
 * Autor: Pablo Sampaio
 */

#include <stdio.h>
#include <stdlib.h>
#include "lexer.h"

// guarda o token atual, lido do lexer
token current_token;

// protótipo das funções
void accept_token();
void accept_token(token_type tp);
void parse_program();
void parse_comand();
void parse_expr();
void parse_restExpr();
void parse_term();
void parse_restTerm();
void parse_factor();


// função principal, a partir da qual é feito reconhecimento
// de um arquivo com código fonte de XPR.
void parse(const char* arquivo) {
     FILE* file = fopen(arquivo, "r");
     
     if (file == NULL) {
         printf(" Erro ao abrir o arquivo: \"%s\".\n", arquivo);
         system("PAUSE");
         exit(1);
     }

	// reinicia o lexer e lê o primeiro token
	reset_lexer(file);
	current_token = next_token();
		
	// tenta reconhecer algo que case com o símbolo "program"
	parse_program();
	accept_token(TOKEN_EOF);
		
	// se nao der erro antes de chegar aqui, então o programa
	// está sintaticamente correto
	printf(" Sintaxe OK!\n");
		
}

///////////////////////////////////////////////////////////
//////////// FUNCOES PARA MANIPULAR OS TOKENS /////////////

void accept_token() {
	current_token = next_token();		
}

void accept_token(token_type tp) {
	if (current_token.type == tp) {
		current_token = next_token();

	} else {
         printf(" ERRO SINTATICO: token inesperado - foi lido um ");
         printf("\"tipo_%d\", quando o esperado era um \"tipo_%d\".\n", current_token.type, tp);
         system("PAUSE");
         exit(1);
	}
}
	
///////////////////////////////////////////////////////////
///////////// METODOS DERIVADOS DA GRAMATICA //////////////
	
/**
 *   <program> ::= <comand>
 */
void parse_program() {
	parse_comand();
}
	
/**
 *   <comand> ::= <expr> ";"
 */
void parse_comand() {
	parse_expr();
	accept_token(TOKEN_PT_VIRG);
}
	
/**
 *   <expr> ::= <term> <restExpr>
 */
void parse_expr() {
	parse_term();
	parse_restExpr();
}

/**
 *   <restExpr> ::= "+" <term> <restExpr>
 *                | --vazio--
 */
void parse_restExpr() {
	if (current_token.type == TOKEN_MAIS) {
		accept_token();
		parse_term();
		parse_restExpr();
	} else {
		/* faz nada */
	}
}
	
/**
 *   <term> ::= <factor> <restTerm>
 */
void parse_term() {
	parse_factor();
	parse_restTerm();
}

/**
 *   <restTerm> ::= "*" <factor> <restTerm>
 *                | --vazio--
 */
void parse_restTerm() {
	if (current_token.type == TOKEN_VEZES) {
		accept_token();
		parse_factor();
		parse_restTerm();
	} else {
		/* faz nada */
	}
}

/**
 *   <factor> ::= NUMERO
 *              | "(" <expr> ")"
 */
void parse_factor() {
	if (current_token.type == TOKEN_NUMERO) {
		accept_token();

	} else if (current_token.type == TOKEN_ABRE_PAR) {
		accept_token();
		parse_expr();
		accept_token(TOKEN_FECHA_PAR);

	} else {
		printf(" ERRO SINTATICO: token inesperado - \"tipo_%d\".\n", current_token.type);
        system("PAUSE");
        exit(1);
	}

}

/////////////////////////////////////////////////////////////

File: XPR/XPR-mix/src/xpr/version1/Parser.java
package xpr.version1;

import java.io.ByteArrayInputStream;
import java.util.HashMap;
import java.util.Map;

import xpr.CompilerException;
import xpr.general.Token;
import xpr.general.TokenType;
import xpr.general.XprLexer;
import xpr.general.XprParser;


/**
 * Esta classe faz o reconhecimento sintático (parsing) da linguagem XPR
 * para representar expressões algébricas simples. Ela calcula o resultado
 * da expressão algébrica à medida que faz o seu reconhecimento.
 * 
 * @author Pablo Sampaio
 */
public class Parser implements XprParser {
	private XprLexer lexer;
	private Token currentToken;
	private Map<String,Integer> symbolTable; 
	
	private StringBuilder output;
	
	public Parser() {
	}

	/**
	 * Metodo principal, a partir do qual é feito reconhecimento
	 * de um dado código fonte em XPR.
	 */
	public String parse(String sourceCode) throws CompilerException {
		output = new StringBuilder();
		symbolTable = new HashMap<String,Integer>();

		// reinicia o lexer e lê o primeiro token
		lexer = new Lexer(new ByteArrayInputStream(sourceCode.getBytes()));
		acceptToken();
		
		// tenta reconhecer algo que case com o símbolo "program",
		// que eh o símbolo inicial da gramática
		this.parseProgram();
		
		return output.toString();
	}

	///////////////////////////////////////////////////////////
	//////////// FUNCOES PARA MANIPULAR OS TOKENS /////////////

	private void acceptToken() throws CompilerException {
		try {
			currentToken = lexer.nextToken();
		} catch (Exception e) {
			throw new CompilerException("Erro inesperado: " + e.getMessage());
		}
	}

	private void acceptToken(TokenType tp) throws CompilerException {
		if (currentToken.getType() == tp) {
			acceptToken();

		} else {
			throw new CompilerException("Token inesperado: "
					        + "foi lido um \"" + currentToken.getType() 
					        + "\", quando o esperado era \"" + tp + "\".");
		}

	}
	
	///////////////////////////////////////////////////////////
	///////////// METODOS DERIVADOS DA GRAMATICA //////////////
	
	/**
	 *   <program> ::= <comand> ( <comand> )*
	 */
	private void parseProgram() throws CompilerException {
		do {
			this.parseComand();
		} while (currentToken.getType() != TokenType.EOF);
	}
	
	/**
	 *   <comand> ::= <definition> ";"
	 *              | <expr> ";"
	 */
	private void parseComand() throws CompilerException {
		String id;
		int valor;
		
		if (currentToken.getType() == TokenType.DEF) {
			id = this.parseDefinition();
			this.acceptToken(TokenType.PT_VIRG);
			
			output.append("  >> definido: " + id + "=" + this.symbolTable.get(id) + "\n");
		
		} else {
			valor = this.parseExpr();
			this.acceptToken(TokenType.PT_VIRG);
			
			output.append("  >> valor: " + valor + "\n");
		}
		
	}

	/**
	 *   <definition> ::= "def" IDENTIFICADOR "=" <expr>
	 */
	private String parseDefinition() throws CompilerException {
		String id;
		int valor;
		
		// neste ponto, o token DEF ja foi reconhecido,
		// então o token atual pode ser aceito
		this.acceptToken();

		// testa se o próximo token é um identificador e lê o seu lexema
		if (currentToken.getType() == TokenType.IDENTIFICADOR) {
			id = currentToken.getLexeme();
			this.acceptToken();
		} else {
			throw new CompilerException("Erro na definição. Foi lido um \"" 
					+ currentToken.getType() + "\", quando o esperado era um identificador.");
		}
		
		this.acceptToken(TokenType.ATRIBUI);
		
		valor = this.parseExpr();
		
		// registra o identificador se ele ainda não existir
		if (! this.symbolTable.containsKey(id)) {
			this.symbolTable.put(id, valor);
		} else {
			throw new CompilerException("Identificador ja existe: " + id + ".");
		}
		
		return id;
	}
	
	/**
	 *   <expr> ::= <term> <restExpr>
	 */
	private int parseExpr() throws CompilerException {
		int resultado;
		
		resultado = this.parseTerm();
		resultado = this.parseRestExpr(resultado);
		
		return resultado;
	}

	/**
	 *   <restExpr> ::= "+" <term> <restExpr>
	 *                | "-" <term> <restExpr>
	 *                | --vazio--
	 */
	private int parseRestExpr(int valorEsquerda) throws CompilerException {
		int resultado = valorEsquerda;
		
		if (currentToken.getType() == TokenType.MAIS) {
			this.acceptToken();
			resultado += this.parseTerm();
			resultado  = this.parseRestExpr(resultado);
		
		} else if (currentToken.getType() == TokenType.MENOS) {
			this.acceptToken();
			resultado -= this.parseTerm();
			resultado  = this.parseRestExpr(resultado);

		} else {
			/* faz nada */
		}

		return resultado;
	}
	
	/**
	 *   <term> ::= <factor> <restTerm>
	 */
	private int parseTerm() throws CompilerException {
		int resultado;
		
		resultado = this.parseFactor();
		resultado = this.parseRestTerm(resultado);
		
		return resultado;
	}

	/**
	 *   <restTerm> ::= "*" <factor> <restTerm>
	 *                | "/" <factor> <restTerm>
	 *                | --vazio--
	 */
	private int parseRestTerm(int valorEsquerda) throws CompilerException {
		int resultado = valorEsquerda;
		
		if (currentToken.getType() == TokenType.VEZES) {
			this.acceptToken();
			resultado *= this.parseFactor();
			resultado  = this.parseRestTerm(resultado);
		
		} else if (currentToken.getType() == TokenType.DIVIDE) {
			this.acceptToken();
			resultado /= this.parseFactor();
			resultado  = this.parseRestTerm(resultado);
		
		} else {
			/* faz nada */
		}
		
		return resultado;
	}

	/**
	 *   <factor> ::= NUMERO
	 *              | IDENTIFICADOR
	 *              | "(" <expr> ")"
	 */
	private int parseFactor() throws CompilerException {
		int resultado;
		String lexema = currentToken.getLexeme();
		
		if (currentToken.getType() == TokenType.NUMERO) {
			resultado = Integer.parseInt(lexema);
			this.acceptToken();
		
		} else if (currentToken.getType() == TokenType.IDENTIFICADOR) {
			if (this.symbolTable.containsKey(lexema)) {
				resultado = this.symbolTable.get(lexema);
				this.acceptToken();
			} else {
				throw new CompilerException("Identificador nao definido: " + lexema + ".");
			}

		} else if (currentToken.getType() == TokenType.ABRE_PAR) {
			this.acceptToken();
			resultado = this.parseExpr();
			this.acceptToken(TokenType.FECHA_PAR);

		} else {
			throw new CompilerException("Token inesperado: " + currentToken.getType() + ".");
		}

		return resultado;
	}
	
	/////////////////////////////////////////////////////////////

	/**
	 * Este metodo executa um pequeno teste de reconhecimento.  
	 */
	public static void main(String[] args) throws Exception {
		Parser parser = new Parser();
		
		String teste = "def x = 22 -4*5 /2; def y = 5; x +y; x * 20; //comentario";
		
		System.out.println(" == TESTE DO PARSER ==");
		System.out.println(" Texto fonte a ser reconhecido: \"" + teste + "\"");
		
		String msg = parser.parse(teste);
		
		System.out.println(msg);
		
		System.out.println(" == FIM ==");
	}

}